main/src/cgeo/geocaching/activity/OAuthAuthorizationActivity.java

```diff
@@ -63,47 +64,67 @@
     private ProgressDialog requestTokenDialog = null;
     private ProgressDialog changeTokensDialog = null;
 
-    private final Handler requestTokenHandler = new Handler() {
+    private final Handler requestTokenHandler = new RequestTokenHandler(this);
+    private final Handler changeTokensHandler = new ChangeTokensHandler(this);
+
+    private static final class RequestTokenHandler extends Handler {
+        private final WeakReference<OAuthAuthorizationActivity> activityRef;
+
+        RequestTokenHandler(final OAuthAuthorizationActivity activity) {
+            activityRef = new WeakReference<>(activity);
+        }
 
         @Override
         public void handleMessage(final Message msg) {
-            if (requestTokenDialog != null && requestTokenDialog.isShowing()) {
-                requestTokenDialog.dismiss();
-            }
-
-            startButton.setOnClickListener(new StartListener());
-            startButton.setEnabled(true);
+            final OAuthAuthorizationActivity activity = activityRef.get();
+            if (activity != null) {
+                final ProgressDialog requestTokenDialog = activity.requestTokenDialog;
+                if (requestTokenDialog != null && requestTokenDialog.isShowing()) {
+                    requestTokenDialog.dismiss();
+                }
 
-            if (msg.what == STATUS_SUCCESS) {
-                startButton.setText(getAuthAgain());
-            } else if (msg.what == STATUS_ERROR_EXT_MSG) {
-                String errMsg = getErrAuthInitialize();
-                errMsg += msg.obj != null ? "\n" + msg.obj.toString() : "";
-                showToast(errMsg);
-                startButton.setText(getAuthStart());
-            } else {
-                showToast(getErrAuthInitialize());
-                startButton.setText(getAuthStart());
+                final Button startButton = activity.startButton;
+                startButton.setOnClickListener(new StartListener(activity));
+                startButton.setEnabled(true);
+
+                if (msg.what == STATUS_SUCCESS) {
+                    startButton.setText(activity.getAuthAgain());
+                } else if (msg.what == STATUS_ERROR_EXT_MSG) {
+                    String errMsg = activity.getErrAuthInitialize();
+                    errMsg += msg.obj != null ? "\n" + msg.obj.toString() : "";
+                    activity.showToast(errMsg);
+                    startButton.setText(activity.getAuthStart());
+                } else {
+                    activity.showToast(activity.getErrAuthInitialize());
+                    startButton.setText(activity.getAuthStart());
+                }
             }
         }
+    }
 
-    };
+    private static final class ChangeTokensHandler extends Handler {
+        private final WeakReference<OAuthAuthorizationActivity> activityRef;
```

Comment:
Could you update `WeakReferenceHandler` so that it accepts an arbitrary type? There is no good reason to restrict it to an activity AFAICS.

You can rename `getActivity()` to `getReference()` to reflect the type change.